### ISSUE TITLE
Where is recenter? GvrCardboardHelpers.Recenter() crashes when VRSettings.enabled=false

### ISSUE THREAD

**Issue by savashito** [googlevr/gvr-unity-sdk#667]
In the previous versions of Cardboard, one could use :

`GVRViewer.Recenter()`
to recenter the camera in VR mode, but now that GVRViewer doesn't exist what is one suppose to do? There is no documentation of the GvrControllerInput in the [https://developers.google.com/vr/unity/reference/class](https://developers.google.com/vr/unity/reference/class).
How do I recenter the camera in VR mode?

**Comment by reed-alpert** [comment]
Typically the app does not recenter, that is left to the user via the Controller Home button.
There is a script in Daydream Elements (ReorientOnLoad) that will face the user towards 'front' at startup.

**Comment by savashito** [comment]
I am not using daydeam, I am using the cardboard viewer, which means no home button. I need recenter because I am creating an app for rowing on a rowing machine, the problem is that one starts centered but after rowing for 5 min and turning your head, the center of the VR boat becomes unaligned with the center of the rowing machine. Before I would simple use  GVRViewer.Recenter(), and everything was nice. Now what should I do to recenter?

**Comment by WeaselKing** [comment]
It works basically the same Savashito. Instead of using that method in GVRViewer it has been moved to GvrCardboardHelpers. Simply do this to recenter:

```
    public void recenterfunction()
    {
        GvrCardboardHelpers.Recenter();
    }
```

I do notice that it crashes the game if you call that function when VRSettings.enabled = false so make sure that can't fire if that is the case.

**Comment by fredsa** [comment]
I can confirm that with GVR  SDK 1.70.0, the app crashes if it tries to call `GvrCardboardHelpers.Recenter()` or `GvrCardboardHelpers.SetViewerProfile()` when the Cardboard VR SDK is not currently activated.

We're working on a fix for an upcoming release. As a workaround you can patch `GvrCardboardHelpers.cs` manually, to make sure that the result of `VRDevice.GetNativePtr()` is not used when it returns `IntPtr.Zero`, which will happen in VR mode is not enabled:

```C#
public class GvrCardboardHelpers {

  public static void Recenter() {
#if (UNITY_ANDROID || UNITY_IOS) && !UNITY_EDITOR
    IntPtr gvrContextPtr = VRDevice.GetNativePtr();
    if (gvrContextPtr == IntPtr.Zero) {
      Debug.LogError("Zero native context pointer");
      return;
    }
    gvr_reset_tracking(gvrContextPtr);
#endif  // (UNITY_ANDROID || UNITY_IOS) && !UNITY_EDITOR
    Debug.Log("Use GvrEditorEmulator for in-editor recentering");
  }

  // …

}
```

Alternatively, you can check the return value in your code:

```C#
void DoRecenter() {
  if (VRDevice.GetNativePtr() == IntPtr.Zero) {
    // Invalid pointer.
    return;
  }

  // Safe, since native context pointer is valid.
  GvrCardboardHelpers.Recenter();
}
```

**Comment by savashito** [comment]
Thanks WeaselKing! I had no idea about the existence of GvrCardboardHelpers, was not available to find that in the documentation.

**Comment by fredsa** [comment]
@savashito Thanks for pointing out the missing docs. I've filed an internal issue to get that fixed.

**Comment by trinquity** [comment]
I still face this problem. If I use  `GvrCardboardHelpers.Recenter()` app crashes on iOS builds. Is there any update about this issue?

**Comment by fredsa** [comment]
For iOS, please also see #621, fixed by the PR in #622.
